contracts/staking/PoSValidatorManager.sol

```diff
@@ -130,4 +144,196 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
+        );
+
+        // Ensure the delegator is not already registered
+        require(
+            $._delegatorStakes[validationID][delegator].weight == 0,
+            "PoSValidatorManager: delegator already registered"
+        );
+
+        _checkAndUpdateChurnTracker(weight);
+
+        // Update the validator weight
+        validator.weight += weight;
+        _setValidator(validationID, validator);
+
+        // Submit the message to the Warp precompile.
+        uint64 nonce = _getAndIncrementNonce(validationID);
+        bytes memory setValidatorWeightPayload = ValidatorMessages
+            .packSetSubnetValidatorWeightMessage(validationID, nonce, validator.weight);
+        $._pendingRegisterDelegatorMessages[validationID][delegator] = setValidatorWeightPayload;
+        bytes32 messageID = WARP_MESSENGER.sendWarpMessage(setValidatorWeightPayload);
+
+        // Store the delegator information
+        $._delegatorStakes[validationID][delegator] = Delegator({
+            status: DelegatorStatus.PendingAdded,
+            weight: weight,
+            startedAt: 0,
+            endedAt: 0,
+            startingNonce: nonce,
+            endingNonce: 0
+        });
+
+        emit DelegatorAdded({
+            validationID: validationID,
+            setWeightMessageID: messageID,
+            delegator: delegator,
+            delegatorWeight: weight,
+            validatorWeight: validator.weight,
+            nonce: nonce
+        });
+    }
+
+    function resendDelegatorRegistration(bytes32 validationID, address delegator) external {
+        _checkPendingRegisterDelegatorMessages(validationID, delegator);
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+        // Submit the message to the Warp precompile.
+        WARP_MESSENGER.sendWarpMessage($._pendingRegisterDelegatorMessages[validationID][delegator]);
+    }
+
+    function completeDelegatorRegistration(uint32 messageIndex, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Unpack the Warp message
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        (bytes32 validationID, uint64 nonce,) =
+            ValidatorMessages.unpackSubnetValidatorWeightUpdateMessage(warpMessage.payload);
+        _checkPendingRegisterDelegatorMessages(validationID, delegator);
+        delete $._pendingRegisterDelegatorMessages[validationID][delegator];
+
+        Validator memory validator = _getValidator(validationID);
+
+        // The received nonce should be no greater than the highest sent nonce
+        require(validator.messageNonce >= nonce, "PoSValidatorManager: invalid nonce");
+        // It should also be greater than or equal to the delegator's starting nonce
+        require(
+            $._delegatorStakes[validationID][delegator].startingNonce <= nonce,
```

Comment:
nvm it's because P-chain is only willing to sign latest nonce weight update message. We keep a start nonce to check for registration completion, and ending nonce for ending completion.